src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotLockOnObjectsWithWeakIdentity.cs

```diff
@@ -47,29 +47,52 @@ public override void Initialize(AnalysisContext analysisContext)
 
             analysisContext.RegisterCompilationStartAction(compilationStartContext =>
             {
-                Compilation compilation = compilationStartContext.Compilation;
+                var compilation = compilationStartContext.Compilation;
+
+                compilationStartContext.RegisterOperationAction(
+                    context => ReportOnWeakIdentityObject(((ILockOperation)context.Operation).LockedValue, context),
+                    OperationKind.Lock);
+
                 compilationStartContext.RegisterOperationAction(context =>
                 {
-                    var lockStatement = (ILockOperation)context.Operation;
+                    var invocationOperation = (IInvocationOperation)context.Operation;
+                    var method = invocationOperation.TargetMethod;
 
-                    if (lockStatement.LockedValue is IInstanceReferenceOperation instanceReference &&
-                        instanceReference.ReferenceKind == InstanceReferenceKind.ContainingTypeInstance)
+                    if ((method.Name != "Enter" && method.Name != "TryEnter") ||
+                        invocationOperation.Arguments.Length == 0)
                     {
-                        context.ReportDiagnostic(lockStatement.LockedValue.Syntax.CreateDiagnostic(Rule, lockStatement.LockedValue.Syntax.ToString()));
+                        return;
                     }
-                    else
+
+                    var monitorType = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingMonitor);
+                    if (monitorType != null &&
+                        method.ContainingType.Equals(monitorType) &&
+                        invocationOperation.Arguments[0]?.Value is IConversionOperation conversionOperation)
```

Comment:
```suggestion
                        invocationOperation.Arguments[0].Value is IConversionOperation conversionOperation)
```

Reply:
As we are receiving an interface object I thought that this could be null so I went for the defensive check :)